### PR TITLE
(CDAP-7033) Perform authorization based on the getDataset access type

### PR DESCRIPTION
**Pull request by chtyim**
- If access type is known, then authorization, lineage and audit will be enforced/recorded
  based on the access type
- If access type is unknown, authorization still need to perform to make sure the user has
  some access right on the entity

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java [hunk 1]
```diff
@@ -107,10 +111,12 @@ protected CallStack childValue(CallStack parentValue) {
                               DatasetAccessRecorder accessRecorder,
                               Principal principal,
                               DatasetId datasetId,
+                              @Nullable Class<? extends Annotation> constructorDefaultAnnotation,
```
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java]: curious about this variable name. why is called constructor'Default'Annotation? Should it just be constructorAnnotation?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java [hunk 2]
```diff
@@ -119,44 +125,45 @@ protected CallStack childValue(CallStack parentValue) {
   }
 
   private DefaultDatasetRuntimeContext(AuthorizationEnforcer enforcer, DatasetAccessRecorder accessRecorder,
-                                       Principal principal, DatasetId datasetId) {
+                                       Principal principal, DatasetId datasetId,
+                                       @Nullable Class<? extends Annotation> constructorDefaultAnnotation) {
```
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:129]: just checking, can there be an annotation on the constructor?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java [hunk 3]
```diff
@@ -119,44 +125,45 @@ protected CallStack childValue(CallStack parentValue) {
...
-    } else {
-      lineageType = callStack.enter(AccessType.UNKNOWN);
-      auditType = AccessType.UNKNOWN;
+    // Performs authorization.
+    // If the method is not annotated, the action set is empty, which means the user need to have some privileges,
+    // but we won't allow no privilege at all
+    try {
+      enforcer.enforce(datasetId, principal, accessInfo.getActions());
+    } catch (Exception e) {
+      throw new DataSetException("The principal " + principal + " is not authorized to access " + datasetId +
```
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:162]: curious, will the exception be complete if the accessInfo.getActions() is empty?

**Comment by gokulavasan** [comment]
One question and one comment. LGTM 👍

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java (hunk 2 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:129]: Yes. If the user annotate it, then the default won't be used.

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-RUT81-2

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java (hunk 3 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java:162]: The exception will shown the action set as an empty set `[]`. But I think it's ok, unless want to specially handle the empty set case. We use empty set to check for if the user has any access right.

**Pull request merged**